### PR TITLE
Display errors and warnings 

### LINKED ISSUE #588: Error handling mechanism
- Load 
  - [ ] Errors
  - [ ] warnings (we can drop something)

- Spec / Plot / Vega-Lite 
   - [ ] For each plot, add logs / warning that are unique to the store 
   - [ ] Whenever the spec changes, clear the log for shelf.

- Filter 
  - [ ] Warn that they cannot drop two filters to the same field. (Show alert)

- ~~Bugs~~ 
  - [x] ~~Undefined Error, etc.  (These are usually disastrous) -- can we design in a way that users can continue.~~

Two options:
1) Distribute error to different parts of the state
    -  Easy to make all other shelf actions clear the error

2) Centralize and have separate keys
    - Easy to combine them in the UI

### PR DESCRIPTION
**Pull request by starry97**
Fix #588

- [x] log UI renders log in the redux store
- [x] Add log in the redux store
- [x] Add localLogger and wire it to plot

### CODE REVIEW

**Comment by starry97** [comment]
screen shot 2017-08-21 at 3 05 02 pm
screen shot 2017-08-21 at 2 46 46 pm

File: src/reducers/shelf/filter.ts [hunk 1]
```diff
@@ -19,7 +19,8 @@ export function filterReducer(shelfSpec: Readonly<ShelfUnitSpec> = DEFAULT_SHELF
       const {filter} = action.payload;
       let index = action.payload.index;
       if (contains(shelfSpec.filters, filter)) {
-        throw new Error('Cannot add more than one filter of the same field');
+        window.alert('Cannot add more than one filter of the same field');
```
Comment (kanitw) [inline src/reducers/shelf/filter.ts]: Hmm, the reducer shouldn't cause side effect to the UI (via `window`). 

Instead, I think throwing error here is totally fine.  

What we need to design is where do we catch these errors and correctly throw error.

**Comment by kanitw** [comment]
Thanks for the PR.  I think this is a good start.  

From looking at code quickly, it seems like there are multiple ways that we handle error.  

After handling issue, could you please summarize what are the mechanism that we have? 
With that I'll be able to give a better higher level comment :)

File: src/actions/dataset.ts [hunk 2]
```diff
@@ -79,7 +80,12 @@ export function datasetLoad(name: string, dataset: Data): DatasetLoad {
           dispatch(ActionCreators.clearHistory());
         });
     } else {
-      throw new Error('dataset load error: dataset type not detected');
+      dispatch({
+        type: LOG_ERRORS_ADD,
+        payload: {
+          errors: ['dataset load error: dataset type not detected']
```
Comment (kanitw) [inline src/actions/dataset.ts]: Dataset load

File: src/reducers/shelf/filter.ts (hunk 1 above)
Comment (kanitw) [inline src/reducers/shelf/filter.ts]: Reading more, throwing error in reducer is not a good idea either.

Instead, we may want to move the validation logic to the relevant filter components

File: src/store/index.ts [hunk 3]
```diff
@@ -28,8 +28,9 @@ if (process.env.NODE_ENV !== 'production' && process.env.NODE_ENV !== 'test') {
   middleware.push(loggerMiddleware);
 }
 
+export const localLogger = new Logger();
```
Comment (kanitw) [inline src/store/index.ts]: I don't think we should have a singleton local logger.  This makes the VegaLite unnecessary coupled.

File: src/components/vega-lite/index.tsx [hunk 4]
```diff
@@ -3,16 +3,19 @@ import * as vega from 'vega';
 import * as vl from 'vega-lite';
 import {TopLevelExtendedSpec} from 'vega-lite/build/src/spec';
 import * as vegaTooltip from 'vega-tooltip';
+import {LOG_ERROR_CHANGE, LOG_WARNINGS_ADD, LogAction} from '../../actions/log';
+import {ActionHandler} from '../../actions/redux-action';
+import {localLogger} from '../../store/index';
```
Comment (kanitw) [inline src/components/vega-lite/index.tsx]: I don't think we should have a singleton local logger.  This makes the VegaLite unnecessary coupled.  Instead, each VegaLite component takes its own logger as input.

File: src/components/vega-lite/index.tsx [hunk 5]
```diff
@@ -3,16 +3,19 @@ import * as vega from 'vega';
...
 
-export interface VegaLiteProps {
+export interface VegaLiteProps extends ActionHandler<LogAction> {
```
Comment (kanitw) [inline src/components/vega-lite/index.tsx]: Update: I would prefer having ~~`warn` and `error` methods as~~
a logger as `VegaLiteProps` and let each logger in `Plot` calls handleAction instead.  (You can name it `plotLogger`)

File: src/models/logger.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
+export class Logger {
```
Comment (kanitw) [inline src/models/logger.ts]: Why do we need this class at all though?  I think we can implement a plotLogger as a part of log, which in turn trigger a relevant LogAction directly.

File: src/reducers/log.test.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
+import {LOG_ERROR_CHANGE, LOG_ERROR_REMOVE, LOG_WARNINGS_ADD, LOG_WARNINGS_REMOVE} from '../actions/log';
+import {DEFAULT_LOG, Log} from '../models/log';
+import {logReducer} from './log';
+
+const simpleLog: Log = {
```
Comment (kanitw) [inline src/reducers/log.test.ts]: Maybe move this down to where it's used?

File: src/components/vega-lite/index.tsx [hunk 8]
```diff
@@ -23,8 +26,37 @@ export class VegaLite extends React.PureComponent<VegaLiteProps, any> {
...
+    //       {"a": "I", "b": 52}
+    //     ]
+    //   },
+    //   "mark": "bar",
+    //   "encoding": {
+    //     "x": {"field": "a", "type": "quantitative"},
+    //     "y": {"field": "b", "type": "quantitative"}
+    //   }
+    // };
+    const {spec} = vl.compile(vlSpec, localLogger);
```
Comment (kanitw) [inline src/components/vega-lite/index.tsx]: Wrap this with try catch

**Comment by starry97** [comment]
Open #650 instead

**Pull request closed without merging**